### PR TITLE
Re-order the shutdown hook sequence

### PR DESCRIPTION
**Pull request by tajila**
Re-order the shutdown hook sequence

The new implementation of Thread.getAllThreads() makes use of method
references. This requires that a new MethodHandleNatives$CallSiteContext
is created. This registers a PhantomCleanable with the commoncleaner and
circumvents what we are attempting to achieve with
CleanerShutdown.shutdownCleaner(). We can resolve this issue re-ordering
operations such that shutdownCleaner is the last operation before we
fallback to the sending the interrupt.

Fixes: https://github.com/eclipse-openj9/openj9/issues/15453

Related: https://github.com/eclipse-openj9/openj9/issues/15408

Signed-off-by: Tobi Ajila <atobia@ca.ibm.com>

### CODE REVIEW

**Comment by tajila** [comment]
jenkins test sanity xlinux jdk19

**Comment by tajila** [comment]
10x Grinder passed 10/10

**Comment by tajila** [comment]
@keithc-ca Ive made the requested changes

**Comment by keithc-ca** [comment]
After looking more closely at this, I have a number of questions/comments:
* How do we know that the `Common-Cleaner` thread has actually been created before the shutdown hook runs? If it hasn't, and something triggers its creation within `shutdownCleaner()`, it won't be among the threads enumerated earlier and so it can't be stopped.
* Have you tested this change with earlier versions (11, 17 & 18)?
* This shutdown code is rather fragile; it may be broken by an application creating a thread named "Common-Cleaner" or a thread group named "InnocuousThreadGroup".
* If the "InnocuousThreadGroup" is not found, a NullPointerException will be thrown.

**Comment by tajila** [comment]
> How do we know that the Common-Cleaner thread has actually been created before the shutdown hook runs? 

Common-cleaner is created as part of JCL initialization, so it will always exist in a successful `JNI_CreateJavaVM` and it will exist very early (pre-main). If startup is not successful, then we dont bother calling `DestroyJavaVM` which is what triggers the shutdownhooks.

> This shutdown code is rather fragile; it may be broken by an application creating a thread named "Common-Cleaner" or a thread group named "InnocuousThreadGroup".
> If the "InnocuousThreadGroup" is not found, a NullPointerException will be thrown.

Likewise "InnocuousThreadGroup" will always exist as its also created in JCL initialization. It is possible for a user to create a new threadGroup with the same name as an existing one, so I'll modify that behaviour.

> Have you tested this change with earlier versions (11, 17 & 18)?

We have existing tests that rely on this behaviour to function properly, memleaks, thrstatetest, gcRegressionTests, etc. All of these have been passing until jdk19.

**Comment by tajila** [comment]
jenkins test sanity xlinux jdk11

File: jcl/src/java.base/share/classes/java/lang/J9VMInternals.java [hunk 1]
```diff
@@ -46,6 +46,7 @@
 import jdk.internal.ref.CleanerShutdown;
 import jdk.internal.ref.CleanerImpl;
 import jdk.internal.misc.Unsafe;
+import jdk.internal.misc.InnocuousThread;
```
Comment (tajila) [inline jcl/src/java.base/share/classes/java/lang/J9VMInternals.java:49]: ill fix the ordering of imports

**Comment by tajila** [comment]
jdk11 PR build - https://openj9-jenkins.osuosl.org/view/Pull%20Requests/job/PullRequest-OpenJ9/2402/

**Comment by tajila** [comment]
thrstatetest grinder on jdk11: https://openj9-jenkins.osuosl.org/job/Grinder/1111/

**Comment by tajila** [comment]
jenkins test sanity xlinux jdk17,jdk18

**Comment by tajila** [comment]
jenkins test sanity xlinux jdk11

**Comment by tajila** [comment]
jdk11 grinder 10/10: https://openj9-jenkins.osuosl.org/job/Grinder/1112/tapResults/

**Comment by tajila** [comment]
jenkins test sanity xlinux jdk17,jdk18

**Comment by tajila** [comment]
jdk 17 grinder 10/10 https://openj9-jenkins.osuosl.org/job/Grinder/1113/
jdk 18 grinder 10/10 https://openj9-jenkins.osuosl.org/job/Grinder/1114/

File: jcl/src/java.base/share/classes/java/lang/J9VMInternals.java [hunk 2]
```diff
@@ -120,40 +121,45 @@ private static void completeInitialization() {
...
 				/*[IF JAVA_SPEC_VERSION < 19]*/
-				ThreadGroup threadGroup = Thread.currentThread().group; // the system ThreadGroup
-				ThreadGroup threadGroups[] = new ThreadGroup[threadGroup.numGroups];
-				threadGroup.enumerate(threadGroups, false); /* non-recursive enumeration */
-				for (ThreadGroup tg : threadGroups) {
-					if ("InnocuousThreadGroup".equals(tg.getName())) { //$NON-NLS-1$
-						threads = new Thread[tg.numThreads];
-						tg.enumerate(threads, false);
-					}
+				ThreadGroup threadGroup = null;
```
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/java/lang/J9VMInternals.java:126]: Please remove the redundant initialization.

File: jcl/src/java.base/share/classes/java/lang/J9VMInternals.java [hunk 3]
```diff
@@ -120,40 +121,45 @@ private static void completeInitialization() {
...
+				try {
+					Field innocuousThreadGroupField = InnocuousThread.class.getDeclaredField("INNOCUOUSTHREADGROUP"); //$NON-NLS-1$
+					innocuousThreadGroupField.setAccessible(true);
+					threadGroup = (ThreadGroup)innocuousThreadGroupField.get(null);
+				} catch (ReflectiveOperationException e) {
```
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/java/lang/J9VMInternals.java:131]: Can you sharpen this catch clause, please?
```
	} catch (IllegalAccessException | NoSuchFieldException e) {
```

File: jcl/src/java.base/share/classes/java/lang/J9VMInternals.java [hunk 4]
```diff
@@ -120,40 +121,45 @@ private static void completeInitialization() {
 
 		if (Boolean.getBoolean("ibm.java9.forceCommonCleanerShutdown")) { //$NON-NLS-1$
 			Runnable runnable = () -> {
-				CleanerShutdown.shutdownCleaner();
 				Thread threads[] = null;
```
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/java/lang/J9VMInternals.java:124]: Please use normal array declaration syntax and remove the redundant initialization:
```
    Thread[] threads;
```

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by keithc-ca** [comment]
jenkins test sanity xlinux jdk11,jdk19

**Comment by tajila** [comment]
The failure is:
```
FAILED: testServerUnreachableForAWhile
java.lang.AssertionError: Failed to properly start server, it terminated prematurely with exit value: 1
	at org.testng.AssertJUnit.fail(AssertJUnit.java:59)
	at jit.test.jitserver.JITServerTest.startProcess(JITServerTest.java:271)
	at jit.test.jitserver.JITServerTest.testServerUnreachableForAWhile(JITServerTest.java:407)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:580)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:716)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:988)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
	at org.testng.TestRunner.privateRun(TestRunner.java:648)
	at org.testng.TestRunner.run(TestRunner.java:505)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:455)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:450)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:415)
	at org.testng.SuiteRunner.run(SuiteRunner.java:364)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:84)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1208)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1137)
	at org.testng.TestNG.runSuites(TestNG.java:1049)
	at org.testng.TestNG.run(TestNG.java:1017)
	at org.testng.TestNG.privateMain(TestNG.java:1354)
	at org.testng.TestNG.main(TestNG.java:1323)
```

I dont think its related to this change, https://github.com/eclipse-openj9/openj9/issues/9197 looks similar

Running grinder https://openj9-jenkins.osuosl.org/job/Grinder/1129/

**Review by keithc-ca — APPROVED** [review]

**Pull request merged**